### ISSUE TITLE
Interested in PR for responsive layout support in RNE?

### ISSUE THREAD

**Issue by idibidiart** [react-native-training/react-native-elements#385]
Hi,

I've just finalized the base features and perf optimizations for https://github.com/idibidiart/react-native-responsive-grid

I've also taken my first PR (better formatted docs)

Are you interested in a PR that ports RNE to use this grid?

I haven't looked at the source for RNE but I'm sure I can tackle it.

Let me know and thanks for making this amazing library. 
:)

**Comment by Monte9** [comment]
@idibidiart Thanks for reaching out! I looked at `react-native-responsive-grid` and it looks great. 💯 

Yes, we are always open to more open source contributions, especially from our users that love our library was wanna contribute back to make it even better.

I haven't looked at the source code for `react-native-responsive-grid`, so can we start by you briefly explaing how it works (high-level) and possibly what an integration into RNE core would look like? Let's use this issue as a discussion thread and then make a PR at the end of it. 

/cc @binoy14 

Also, we have a slack team where you can come hang out with the RNE community. https://reactnativetraining.herokuapp.com/. See you on the other side! :)

**Comment by idibidiart** [comment]
Hey @monte9! :)

So glad you're open to this.

I feel that the latest Readme and design update makes the What, Why and How of this really obvious.

The goal is to make RNE responsive at the component level such that both the existing components and more complex/rich components can respond and adapt to screen size, aspect ratio and orientation change in such a way that the RNE user would be able to easily define aspects of the responsive/adaptive behavior if they wish to (use custom rules) To do that, we have to have a simple, clean, and declarative way to specify responsive and adaptive behavior. This grid is that.

Here is a copy/paste from the updated Readme that describes the grid itself:


## Background

Priot to React Native v0.42 there was no performant and declarative way of specifying percentage-based style values. Then came React Native v0.42 with performant support for percentage-based dimensions that may be declared in the style prop. Since then several open source contributors have made responsive grids that take advantage of this new capability. I've taken one of the simplest and most well-thought-out ones, namely, `react-native-flexbox-grid` (by @rundmt), and made some significabnt changes and enhancements to it that have resulted in a simple and powerful layout model with predictable dynamic behavior. 

## Design TL;DR

This grid eschews Flexbox-based sizing in favor of percentage-based sizing while at the sane time relying on Flexbox for alignment. It delivers the predictability of the percentage-based model with the more sophisticated alignment capasbility of the Flexbox model.

_The "grid" construct is extended beyond its archetypal form as a regular 2D mesh and generalized as a nestable pattern of Rows containing one or more Columns and Columns containing one or more Rows. Both Rows and Columns can be styled using predictable percentage-based dimensions, padding and margins. Rows can be aligned vertically inside Columns and Columns can be aligned horizontally inside rows. [Additionally, the multiple lines created by wrapped Row content may be aligned vertically relative to the parent Column.] This gives us the ability to build the entire UI component tree (or an individual component's subtree) using a consistent, repeatable and nestable pattern, one that has dynamic --not only static-- responsiveness and row-level layout events (for adaptive UI) built into it._

(for more details see [Design Principles](https://github.com/idibidiart/react-native-responsive-grid/blob/master/DesignPrinciples.md))

## Examples

You may use this grid to build responsive 2D layouts that maintain their relative proportions, change their basic structure in a predictable way and dynamically decide what content to display, based on screen size, aspect ratio, and orientation.  

## [Example 1: aspectRatio](https://www.youtube.com/watch?v=Nghqc5QFln8)
[![>> aspectRatio demo <<](https://img.youtube.com/vi/Nghqc5QFln8/0.jpg)](https://www.youtube.com/watch?v=Nghqc5QFln8)

## [Example 2: breakPoints](https://www.youtube.com/watch?v=GZ1uxWEVAuQ) 
[![>> responsive break points demo <<](https://img.youtube.com/vi/GZ1uxWEVAuQ/0.jpg)](https://www.youtube.com/watch?v=GZ1uxWEVAuQ)

## [Example 3: layoutEvent](https://www.youtube.com/watch?v=99J3c_Zn6QU) 
[![>> layoutEvent demo <<](https://img.youtube.com/vi/99J3c_Zn6QU/0.jpg)](https://www.youtube.com/watch?v=99J3c_Zn6QU)

## [Example 4: FlatList](https://www.youtube.com/watch?v=qLqxat3wX_8)
[![>> FlatList Demo <<](https://img.youtube.com/vi/qLqxat3wX_8/0.jpg)](https://www.youtube.com/watch?v=qLqxat3wX_8)

The demos in the videos above show some of the possibilities, but this grid is capable of more complex responsive and adaptive behavior.

**Comment by idibidiart** [comment]
Alternatively, Row and Column could be the two new composable components that we'd add to RNE.

I didn't go with a <Grid> component because 1) it's an extra component that is not needed and 2) it doesn't work to have a Grid per item in something like FlatList. Instead, I have separate nestable Row and Column components where Row is the responsive container and Column is the child in that container (that can have a Row as child and so on, as the nesting goes) All Rows listen to each other's layout changes (which is how dynamic response is imolemented) or broadcast their change event to whichever component is interested (that is for any custom behavior the user may wish to define; responsiveness is automatic and by design) 

It would seem to me that just introducing the composable Row/Column pattern into RNE will allow the users to build responsive components, and if that works out well the next step could be to rewrite the existing components using this composable, responsive, devlarative pattern.

Your thoughts? :)

**Comment by idibidiart** [comment]
Typo: should say: I didn't go with a literal *Grid* component

**Comment by binoy14** [comment]
@idibidiart this looks great. Great work on the library we will definitely be interested in adding this to rne but my only concern is compatibility with older versions of react native. We have lot of people who are using an older version of react native and we definitely need to maintain that compatibility. Maybe we create a separate component such as ResponseGrid which will live alongside the existing grid. What do you think?

**Comment by idibidiart** [comment]
@binoy14 

The only reason it depends on RN 0.43+ is to support FlatList and alignContent. But I do agree with you that it could be a special branch. It is also fairly new and while I've been using it in all my React Native projects and experiments it still needs a bigger user base to bang on it so we know that it works well for all kinds of users. 

Only one note: 

I did not create a single "Grid" component because having the Row do the job of a container rather than adding an extra "Grid" component for that role made sense, especially when you have to use FlatList or SectionList where you have to provide just a Row node with its subtree and nothing more, as opposed to having to create a Grid per FlatList row or wrap FlatList in a Grid, both of which are sub-optimal. In fact having a literal Grid component can be very confusing when you're trying to compose a tree structure. Take the following issue from NativeBase' Easy Grid repo. Someone is suggesting a sub-optimal solution here:

https://github.com/GeekyAnts/react-native-easy-grid/issues/31#issuecomment-302691165
```
roffelund <notifications@github.com>
5:36 AM (2 hours ago)

to GeekyAnts/reac., Subscribed 

A quick fix is nesting a new Grid in the row in or col.
<Grid> <Row> <Grid> <Row> </Row> <Row> </Row> </Grid> </Row> <Row> <Grid> <Row> </Row> <Row> </Row> </Grid> </Row> </Grid>
```

That's terrible IMO.

So instead what I have are just Row and Column components that an compose to build a responsive/adaptive UI tree structure.

Here is the bit from the just updated Readme:

```
Design TL;DR

This grid eschews Flexbox-based sizing in favor of percentage-based sizing while at the sane time relying on Flexbox for alignment. It delivers the predictability of the percentage-based model with the more sophisticated alignment capasbility of the Flexbox model.

The "grid" construct is extended beyond its archetypal form as a 2D Rectilinear Grid and generalized as a pair of components, namely, Row and Column, that can be used to compose a tree of UI Views. That is to say, a Row View (Row for short) may contain one or more Column Views (Column for short) each of which may contain one or more Rows, which in turn may contain one or more Columns, and so on. Both Rows and Columns can be styled using predictable, percentage-based width, height, padding and margins. Rows can be aligned vertically (along main axis) and horizontally (along cross axis) inside Columns, and Columns can be aligned horizontally (along main axis) and vertically (along cross axis) inside Rows. [Additionally, the multiple lines created by wrapped Row content may be aligned vertically relative to the parent Column.] This gives us the ability to build the entire UI component tree (or an individual component's subtree) using a consistent, repeatable and nestable pattern, one that has dynamic --not only static-- responsiveness and adaptivity built into it.

(for more details see [Design Principles](https://github.com/idibidiart/react-native-responsive-grid/blob/master/DesignPrinciples.md))

```

So we could add it as separate Row and Column components to RNE that are used in conjunction to build responsive/adaptive UI. 

I need to test that proposal first and make a demo using existing RNE elements to show how that would work.

It could also be in an experimental branch of RNE.

What do you guys think?

**Comment by idibidiart** [comment]
As they say, wording is everything. It's hard sometimes to articulate a new paradigm because we tend to have inertial bias toward the paradigm we're used to. So I've made a better attempt at defining the essence of the "grid" in a way that leaves no room for ambiguity as to how it should be used:

"The "grid" construct is extended beyond its archetypal form as a Rectilinear 2D Grid and generalized as a cross-nestable pair of components, namely, Row and Column, that can be combined in an alternating pattern to build a tree of Views of any shape and depth. That is to say, a Row View (Row for short) may contain one or more Column Views (Column for short), each of which may contain one or more Rows, which in turn may contain one or more Columns, and so on. Both Rows and Columns can be styled using predictable, percentage-based width, height, padding and margins. Rows can be aligned vertically (along main axis) and horizontally (along cross axis) inside Columns, and Columns can be aligned horizontally (along main axis) and vertically (along cross axis) inside Rows. [Additionally, the multiple lines created by wrapped Row content may be aligned vertically relative to the parent Column.] This gives us the ability to build the entire UI component tree (or an individual component's subtree) using a consistent, repeatable and nestable pattern, one that has dynamic --not only static-- responsiveness and the ability to trigger specific adaptive behavior built into it."

Hope that clarifies how we'd go about integrating it into RNE. Basically, I propose to create an experimental bramch of RNE that requires RN 0.43+ and has two new components: Row and Column. RNE users will be able to use them to build their own responsive/adaptive components. If that experiment works out well then it is possible to rewrite exing RNE components using these two core components, to get responsiveness and adaptive behavior baked into RNE in a way that can be easily manipulated (at JSX level) by the users of the library.

If that makes sense then great! If not, I look forward to your thoughts on it! 

Thank you both for being open to the raw idea. It is a sound one. We just have to think deeply about it, so we make sure it makes total sense.

Thanks

**Comment by camillo777** [comment]
Hi
Just as a curiosity, how do you assure that using an array of rows the columns are then perfectly aligned together?

**Comment by idibidiart** [comment]
@camillo777

In the Readme there is a section on "offset" prop. It works in both left to right (normal) as well as right to left directions.

Giving the same offset value to corresponding columns across rows assures their perfect alignment.

**Comment by idibidiart** [comment]
You don't lose any aspect of the "grid" Just a normalized and consistent way of building responsive UI component trees. A grid is a tree where the rows are the children and the columns within each are the grand children. But UI is constructed as a tree of arbitrary shape that corresponds to the UI's visual composition, so a rectilinear 2D grid is the simplest tree structure but for a layout farmework, which is what the responsive grid is, you want the ability to build UI View trees of any shape, depth and complexity, and still have a predictable responsive behavior. That is what the responsive grid gives you. It's a granular responsive layout, not just a design/formatting tool. You could certainly build a high level grid out of it, but its utility would then diminish. That's why I'm proposing to include it in RNE in its composable granular form.

**Comment by idibidiart** [comment]
"When I first made a grid I happened to be thinking of the innocence of trees and then this grid came into my mind and I thought it represented innocence, and I still do, and so I painted it and then I was satisfied. I thought, this is my vision." --Agnes Martin

I know I can build the most beautiful components and layouts with it and have them be alive, reacting to device orientation, and different device sizes, responsively and adaptively, and I know it is the right level of abstraction, but i have a nagging fear that it's too hard for most users to deal with that which is new and unfamiliar even if it is fundamentally sound as a pattern. 

I might have to make my own component library based on it but only as the need for each new component arises and maybe I can then contribute them to RNE because I really don't want to be in the business of making component libraries, and RNE is great and has you guys behind it.

Maybe we just need more time to explore the possibilities. 

Please keep this open and let's imagine potential outcomes. I know that this granular and consistent layout pattern can make RNE components responsive and adaptive, and people have found it easy to build components with.

I look forward to your thoughts on this. I believe I explained everything at the design/abstract level.

**Comment by idibidiart** [comment]
@Monte9 @binoy14 

Here is a self-contained example of how a responsive view tree for List Item component can constructed using the grid's Row and Column pair. It's taken out of the updated Readme.

As you can see, anyone should be able to compose their own custom responsive components by simply using Row and Column to build the responsive view tree. It is a very simple process (much simpler than using raw Flexbox) once the essential pattern is understood.

The user of this library may also build a high level rectilinear grid component using Row and Column and use it to layout screens, but it will be just a high level grid and the elements themselves won't be responsive. To make the entire view tree of a component responsive the user must use Row and Column, following a consistent, repeatable and nestable pattern.

you could say that this is a new way to do layout where custom components can be built very easily in terms of their layout. If an experimental branch of RNE is rewritten in this grid (Row and Column) then all components can be made responsive and RNE users can edit the JSX to tweak the responsive behavior.

The fear, always, is that may be the wrong level of abstraction for some people, but for serious users who want to build responsive/adaptive layouts and components fast and with predictable behavior and performance there is no easier way than adopting  this grid, or something at a similar level of abstraction.

### Example 4

FlatList is a virtualized replacement for React Native's old ListView component. Using FlatList as a container is supported by this grid. This example also demonstrate wrapping Column content based on screen size. See ('size' prop under the [Props](https://github.com/idibidiart/react-native-responsive-grid#props) section.) It also demonstrates who to wrap Row content (e.g. child columns) based on screen size (also see [Example 2](https://github.com/idibidiart/react-native-responsive-grid#example-2) for more details)

```jsx
import React, { Component } from 'react';
import {
  FlatList,
  Text,
  ScrollView
} from 'react-native';

import { Row, Column as Col} from 'react-native-responsive-grid'
import { MaterialIcons } from '@expo/vector-icons';
import faker from 'faker';

let j = 0
const randomUsers = (count = 10) => {
  const arr = [];
  for (let i = 0; i < count; i++) {
    arr.push({
      key: faker.random.uuid(),
      date: faker.date.weekday(),
      name: faker.name.firstName(),
      job: faker.name.jobTitle(),
      index: j++
    })
  }
  return arr
}

export default class Home extends Component {
  state = {
    refreshing: false,
    data: randomUsers(10),
  };

  onEndReached = () => {
    const data = [
        ...this.state.data,
        ...randomUsers(10),
      ]

    this.setState(state => ({
      data
    }));
  };

  onRefresh = () => {
    this.setState({
      data: randomUsers(10),
    });
  }

  render() {
    return (
      <FlatList
        data={this.state.data}
        initialNumToRender={10}
        onEndReachedThreshold={1}
        onEndReached={this.onEndReached}
        refreshing={this.state.refreshing}
        onRefresh={this.onRefresh}
        renderItem={
          ({ item }) => {
            return (
              <Row key={item.key} style={{paddingTop: '6%', paddingBottom: '6%', backgroundColor: 'white', borderBottomColor: 'lightgray', borderBottomWidth: 1}}>
                <Col size={80} offset={6} >
                  <Row wrap>
                    <Col size={60} >
                      <Text style={{fontSize: 15, color: '#BD1206', fontWeight:'bold'}}>{String(item.date)}</Text>
                      <Row>
                        <Col size={10}>
                          <MaterialIcons name='person' size={17} color='gray'/>
                        </Col>
                        <Col smSize={60} size={87.5} offset={2.5}>
                          <Text style={{fontSize: 12, color: 'gray', lineHeight: 20}}>{item.job}</Text>
                        </Col>
                      </Row>
                    </Col>
                    <Col size={40} breakPoints={{sm: 200}}>
                      <Text style={{fontSize: 16, color: '#0a0a0a'}}>{item.name}</Text>
                    </Col> 
                  </Row>    
                </Col>
                <Col size={8} offset={-6} hAlign='right'>
                      <Text>{item.index}</Text>
                </Col>
              </Row>
            )
          }
      }
    />)
  }
}
```

**Comment by idibidiart** [comment]
Updated description of the layout building model and upgraded layout event handling to fit well into React's layout update model.

The changes are:


## A Unified Theory of Layout

This grid eschews Flexbox-based sizing in favor of percentage-based sizing while at the same time relying on Flexbox for alignment. It delivers the predictability of the percentage-based model with the more sophisticated alignment capasbility of the Flexbox model.

_The grid construct is extended beyond its archetypal form as a Rectilinear 2D Grid and generalized as a cross-nestable pair of components, namely, Row and Column, that can be combined in an alternating pattern to build a tree of Views of any shape and depth. That is to say, a Row View (Row for short) may contain one or more Column Views (Column for short), each of which may contain one or more Rows, which in turn may contain one or more Columns, and so on. Both Rows and Columns can be styled using predictable, percentage-based width, height, padding and margins. Rows can be aligned vertically (along main axis) and horizontally (along cross axis) inside Columns, and Columns can be aligned horizontally (along main axis) and vertically (along cross axis) inside Rows. [Additionally, the multiple lines created by wrapped Row content may be aligned vertically relative to the parent Column.] And with this, we can build the entire UI component tree (or an individual component's subtree) using a consistent, repeatable and nestable pattern, one that has predictable and dynamic --not only static-- responsiveness and the ability to trigger specific adaptive behavior, both built into it._ 

_The tree nodes are represented by Rows and their children are represented by Columns. Tree leaves are Columns that have no Row as a child._

### A grid or a tree?

With this grid, we don't lose any aspect of the normal rectilinear 2D grid. We just gain a generalized and consistent way of building responsive UI component trees. A normal rectilinear 2D grid is a tree where the rows are the children and the columns within each are the grand children. But a UI is constructed as a tree of arbitrary shape which corresponds to the UI's visual composition, so a normal rectilinear 2D grid is the simplest tree structure but for a layout farmework, which is what the responsive grid is, we want the ability to build UI View trees of any shape and complexity, and still have predictable responsive behavior. That is what this responsive grid gives us. It's a granular responsive layout, not just a design/formatting tool. We could certainly build a high level normal rectilinear 2D grid out of it, but its utility would then diminish.

(for more details see [Design Principles](https://github.com/idibidiart/react-native-responsive-grid/blob/master/DesignPrinciples.md))


And here:


### Example 3

In the third demo, normally, each Row automatically rerenders itself upon receiving React Native's layout change event after running any pending transitions (library we do not have to do anything.) If 'layoutEvent' is supplied as a prop on a Row with a Row-specific event name the Row will emit that event upon receiving React Native's layout change event rather than re-render. This is useful when we wish to react to layout change on per-row basis. The example below shows how we may listen and react to such specific layout events in components.

....

No changes in API.

I think that adding the Row and Column components to an experimental branch of RNE would give us a way to evaluate how this layout model will work for the average user. Specifically, will it be used for general layout with existing RNE components or if it will be used mostly by those who wish to build their own responsive/adaptive components? I cannot tell ahead of time how peoplewill perceive and use the Row and Column components if we provide them as is but either way I believe this purist design can gel with RNE once users start using them. There may be other requirements that come up that we're not ware of and can be taken care of then, in that special branch. 

What do you think?

**Comment by Monte9** [comment]
@idibidiart this issue is a gold mine for anyone wanting to understand using a grid layout in react-native. Thanks for all the detailed explanations. 💯 👏 

Also, reading through the above thread, it seems like you clearly know what you are talking about and also I feel like you are capable of leading this project by yourself. 

So looking ahead, I think this is how we should proceed:
1. I would like to see this Grid Layout implemented in on of our components to get a better sense of what his might look like in practice. 
2. Also as @binoy14 pointed out, all of our changes should be backward compatible. So if any breaking changes are made, I would like to know how significant they are and if we can minimize such changes. 

TLDR; 
1. Please take the lead of this project
2. Pick a component that you think might be ideal for the Grid Layout implementation and submit a PR
3. On the PR, briefly, explain the code refactoring and implications for our users when releasing it without next release. 

Again, thanks for all the awesome discussion so far, and sorry for the delayed response.

**Comment by idibidiart** [comment]
Excellent! 

I have some projects in flight but will get to this as soon as I free up. I'll also add support for alignSelf and the baseline option where applicable. I hadn't needed those myself but they'll complete the Flexbox support in the grid. 

Will ping you here with updates on my progress with this task.

So exciting @Monte9  !!! :)

**Comment by idibidiart** [comment]
There is a bug in flexWrap that I'm trying to get some attention on

https://github.com/facebook/react-native/issues/8960#issuecomment-307138295

Just FYI... doesn't affect the grid unless wrap prop is set on Row and only in a certain edge case... 

I'll proceed with adding alignSelf and baseline alignment option as those are unrelated but it's important to find out from the RN maintainers if there is a real bug in flexWrap (I think so) so can all use it without worry ...

**Comment by idibidiart** [comment]
Update:

Sorry for the slow-motion progress.

It looks like the bug was in an older version of the Yoga layout engine that RN depends on.

https://github.com/facebook/yoga/issues/575#issuecomment-309307224

I'll test with latest Yoga and if all is good I'll proceed with the wrapping up (pun intended) the remaining gap to support the full Flexbox spec.

**Comment by idibidiart** [comment]
Update:

Just pushed a new release in preparation to rebuilding an RNE component with this grid. I've added a CHANGELOG.md that I'll be maintaining since I figured it's important for others to stay on top of any changes and additions.

Even before this release, I've been sticking to making non-breaking changes only, or providing backward compatibility when a breaking change is necessary (none so far.) 

https://github.com/idibidiart/react-native-responsive-grid/blob/master/CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.29.88 & 0.29.89] - 2017-06-20
### Added
- Points-based `size` and `offset` props since absolute positioning is sometimes needed in a responsive layout, e.g. position of back arrow and hamburger menu on navigation bar. 
- `baseline` value for `vAlign` when supplied to Row
- This CHANGELOG.md

### Changed
- Fixed device-size-based offsets 
- Fixed regression in item alignment for RTL mode 
- Fixed handling of left/right margins (when set in style) rekative to offset value and RTL/LTR modes
- Refactored code for clarity and correctness
- Updated Readme to explain use of newly added features

Will keep updating this thread as I make progress toward the goal. I might take a detour with a feature I've been thinking about that will expand the creative possibilities on larger devices. Then I'll be totally free to dive into the RNE experiment.

Stay tuned! Thanks!

**Comment by idibidiart** [comment]
Hey guys,

This latest examples shows how you can build Universal Apps of consistent UX using nothing but the grid.

This one is about Universal, Responsive Tiles.

https://www.youtube.com/watch?v=x785Qib0ySg

The problem it solves is summarized below:

How to make a tiled screen layout that is highly usable and looks consistent across all screen sizes and aspect ratios, and how to do that using this grid (react-native-responsive-grid.) This involves the following:

1. How to size tiles such that they change size relative to the size of the screen *as well as* retain their shape (width/height aspect ratio)

2. How do we hide/show tiles on demand and fill the void left by hidden tiles using Flexbox wrapping behavior 

The goal is how to do the above in an elegant and declarative way that allows the average user to work without all the tedious implementation details of doing it in row Flexbox and JS. 

Latest version is 0.31.0

Changelog:

https://github.com/idibidiart/react-native-responsive-grid/blob/master/CHANGELOG.md


After this, if there are no bugs, I'll be ready to take on the RNE integration...

**Comment by idibidiart** [comment]
@Monte9 

I finally arrived at what I'm confident is the right approach for me to take this forward.

I'm going to build an open source community bulletin board app using RNE components that are rewritten using the grid, and make that app as a whole work on all device sizes and aspect ratios. Only by such holistic demonstration will anyone actually get the benefit of using the grid :) 

People need something they could touch and feel not an abstract potential. So that's what I'll be doing :)

Marc

**Comment by Monte9** [comment]
Sounds good @idibidiart! If you have an expo app demo for it, please feel free to post the link here, so people can check it out.
